[v1.5][F09-B3] CoinBook

laichengyu opened the pull request:
@ewaldhew @Eldon-Chung  @neilish3re

Commit:
Add COMMAND_ALIAS as an attribute to each command.

Commit:
Modify AddressBookParser to recognise Command aliases and modify test cases to test for Command aliases

Commit:
Update the user guide to mention command aliases

ewaldhew committed:
[#9] Add support for command aliases (#13)

This reduces the amount of typing needed by the user to call the commands.

laichengyu committed:
Add backupAddressBook method to AddressBookStorage interface

laichengyu committed:
Add overloaded backupAddressBook method

laichengyu committed:
Remove unneeded overloaded method

laichengyu committed:
Add backupAddressBook method to StorageManager class

laichengyu committed:
Add backupAddressBook method to XmlAddressBookStorage class

laichengyu committed:
Remove trailing whitespaces

ewaldhew committed:
[#12] Add backup API for address book storage (#17)

- Add empty `backupAddressBook' method to `AddressBookStorage' interface
- Implement `backupAddressBook' method in `XmlAddressBookStorage' class.
- Implement `backupAddressBook' method for `StorageManager' by calling `AddressBookStorage''s backup method

ewaldhew committed:
[#11] Various UI enhancements (#16)

* PersonCard#tags: Add colors to the labels
* NewResultAvailableEvent: Add error flag
* ResultDisplay: Make it actually set error style
* StatusBarFooter: Implement support for item count
* Update all tests

ewaldhew committed:
v1.0 Documentation (#29)

- UserGuide.adoc: Update links
- README.adoc: Update badges
Add acknowledgements to se-edu
- DeveloperGuide.adoc: Update links
Add more NFRs
Add User Stories
Add target user profile
Add glossary
Add Use cases
Fix formatting in Sections 1,4,5,6

- build.gradle: Add trailing whitespace autofix for docs
- Disable `CATASTROPHIC FAILURE: Whitespace errors'

ewaldhew committed:
Update docs images (#32)

- Add UI mockup image
- Add profile photos

neilish3re committed:
[#23] README.adoc: Add product description (#31)

laichengyu committed:
[#25] Add fetch utility class to obtain data from external APIs (#39)

* Add FetchUtil commons class
* Add method to deserialize an array of JSON objects

eldon-chung committed:
[#28] Add news panel (#34)

* NewsCard: Add class and fxml to contain news links
* NewsListPanel: Add class and fxml to contain `NewsCards'
* MainWindow: Add `BrowserWindow' to open links when NewsCards are clicked on

eldon-chung committed:
[#21] UserGuide: Update user guide to describe the project (#38)

ewaldhew committed:
[#22] AboutUs.adoc: Update with own team (#35)

* New fork, New team
* Fix image format.
* Create portfolio files.

ewaldhew committed:
[#27] Charts panel component (#33)

* ChartsPanel: Make a new UI component for drawing charts
* ChartPanel: Add sample data
* Prepare for integration with web API

ewaldhew committed:
v1.1: Fix warnings (#42)

* ChartsPanel: Unchecked assignment to generics
* Fix Checkstyle warnings

ewaldhew committed:
Version 1.1 released (#48)

eldon-chung committed:
[#36] Merge browser window (#47)

* BrowserWindow: Make this handle both help and web pages
* MainWindow#handleHelp(): Use `BrowserWindow' for help page
* Refactor jUnit tests to use `BrowserWindow' instead of `HelpWindow'

k-l-a commented on README.adoc:
```diff
@@ -1,11 +1,9 @@
-https://gitter.im/se-edu/Lobby[image:https://badges.gitter.im/se-edu/Lobby.svg[Gitter chat]]
+https://travis-ci.org/CS2103JAN2018-F09-B3/main[image:https://travis-ci.org/CS2103JAN2018-F09-B3/main.svg?branch=master[Build Status]]
+https://ci.appveyor.com/project/ewaldhew/main[image:https://ci.appveyor.com/api/projects/status/anm4ynat6657reac?svg=true[Build Status]]
+https://coveralls.io/github/CS2103JAN2018-F09-B3/main?branch=master[image:https://coveralls.io/repos/github/CS2103JAN2018-F09-B3/main/badge.svg?branch=master[Coverage Status]]
```
Good to see you've updated these.

k-l-a commented on src/main/java/seedu/address/model/CryptoCompareHistoHourObject.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    /**
+     * Get the price history (per day open) data
```
Should be `Gets` instead of `Get` to be consistent. Maybe use `Returns` to be more precise, instead.

k-l-a reviewed (commented):
Please update the UG to document all features up to V2.0, and mark them accordingly. Update your Readme title to reflect your product name, as well. This was required for last week's team component. Therefore, it has to be considered missed.

Also, please update the title of this PR to reflect your product version each time you release a new version. (e.g. [v1.1])

@laichengyu @Eldon-Chung @ewaldhew @neilish3re 
Furthermore, I would want you to mention the following in your **Developer guide**
1. For each team member what are the features (major and minor) you are proposing?
2. Within the scope of the project, how does it fit in (just a one or 2 line summary)
Please do this by end of day Monday 19 March.

ewaldhew commented:
@k-l-a Re: UG updates, this was done in https://github.com/nus-cs2103-AY1718S2/addressbook-level4/pull/62/commits/674fa0cbad3d9f8e003701c5d035127f791ea20d, with the usage of the new commands we are planning to add, so we're not sure which requirement we missed, would appreciate clarification.

Will update the name of the product in docs, missed the header despite updating the body text (surely this minor oversight isn't why it's considered missed?)

Thanks for the review!

ewaldhew committed:
DeveloperGuide.adoc#App.B: Add Feature Contribution (#66)

ewaldhew committed:
v1.2 Documentation (#64)

k-l-a commented:
@ewaldhew Are all the features there available in the current version? It was required to mark/tag them as coming in v2.0 (or v1.5 or similar) for features that are not yet implemented. Currently I can only see 1 feature marked as coming in v2.0.

If all the unmarked features are indeed available in the current version, then you need to document more than just 1 feature for v2.0. It should describe what your application *should* have by 2.0

ewaldhew committed:
[#41] UsingCheckstyle.adoc: Add crucial step into instructions (#73)

eldon-chung committed:
[#40] UserGuide.adoc: Mark features (#76)

laichengyu committed:
[#40] Add autocomplete to user guide (#75)

laichengyu committed:
[#58][#63] Generic URL builder for API calls (#77)

* Add UrlBuilderUtil
* Update readme to reflect use of Apache HttpClient library

ewaldhew committed:
[#65] v1.2 Documentation: Developer Guide (#67)

* Add new markup style for file links
* DeveloperGuide.adoc: Delete suggested tasks as LOs
Edit for form and clarity
Add section 3.1 to 3.5
Add sequence diagrams

ewaldhew committed:
[#61] Prepare for new feature (#68)

* NotifyCommand: Add basic code
* RuleBook: Create all associated classes empty

eldon-chung committed:
[#26] Refactor person class (#78)

* Refactor all instances of Person to Coin
* Coin.java: Add Amount and Price class to Coin class
Add Amount held, sold, bought and profit calculation
* CoinCard.java: Remove/add fields that are to be displayed on the UI

ewaldhew committed:
Version 1.2 released (#79)

ewaldhew committed:
Documentation: Edit for style, Add diagrams (#86)

ewaldhew committed:
[#26] Change most instances of `Address` to `Coin` (#87)

* Coin: Refactor field: `Phone` to `Code`
Remove field `Address`, `Email` 
* Change `AddressBook` -> `CoinBook`
* AddCommandParserTest: Remove `Address` references

ewaldhew committed:
Relink README, Add UG usage (#100)

eldon-chung committed:
[#46] Add Find command arguments parser (#102)

* conditionalparser: Add Token, TokenStack, TokenType, Lexer and SyntaxParser classes and related tests
* FindCommandParser: Refactor to use Lexer, SyntaxParser

ewaldhew committed:
Change Edit command to Tag command (#98)

* Remake `Code` class
* Change `EditCommand` to `TagCommand`
* Make TagCommand ignore name change

ewaldhew committed:
[#104] BuyCommand: Add new command (#106)

Has limited functionality as of now. The updated data is also not being
written to disk yet, so is not persistent.

* StringUtil: Add validation for decimals.
* ParserUtil: Add double parsing function.
* BuyCommand, BuyCommandParser: Add new files.
* AddressBookParser: Register new command.
* Amount, Coin: Add various copy constructors.

ewaldhew committed:
Version 1.3 released (#107)

laichengyu committed:
[#89] Add SyncCommand (#110)

#89 
* Add syncAll method to Model and ModelManager
* Add GSON dependencies
* Update readme to include GSON library usage
* Add fetchAsJson utility method
* Expose getCodeList API of CoinBook
* Add sync command
* Add SyncCommand
* Add syncAll exceptions
* Add overloaded constructor for Price and Coin class with newPrice
* Add syncAll, getCodeList API method to CoinBook

eldon-chung committed:
[#111] Merge Lexer.java into ArgumentTokenizer.java (#112)

#111 
* TokenType.java: Refactor SPECIFIER type to multiple PREFIXES as needed.
* parser: Move Token, TokenStack, TokenType class to parser package.
* SyntaxParser.java: Add isPrefix method to match prefix type Tokens
* parser: Remove CliSyntax, Prefix class
* ArgumentTokenizer.java: Merge Lexer methods into ArgumentTokenizer.
* TokenType.java: Revise test Tokens to use new PREFIX_Names.
* parser: Refactor CommandParsers to use TokenType when tokenizing args.
* TagCommand.java: Refactor "address" in MESSAGE_USAGE to "coin".
* AddCommand.java: Refactor "address" in MESSAGE_USAGE to "coin".
* ArgumentTokenizerTest.java: Add tests for tokenizeToTokenList method.
* BuyCommandParser.java: Add Parser interface to class definition.
* TokenType.java: Revise regex for Decimal type.
* TokenType.java: Fix order of precedence the regex types are matched.
* BuyCommandParserTest.java: Add unit tests for BuyCommand class.

ewaldhew committed:
Version 1.3b released (#113)

ewaldhew committed:
Add `buy` command related tests (#108)

* Change edit coin constructor to preserve other fields
* StringUtil#isValidNumber: Fix validation
* Coin: Fix values in sold, bought, and price for testing
* Add tests for `buy` command changes

eldon-chung committed:
[#117] Add multi-word tags (#119)

* TokenStack.java: Add whitespace removal during construction and tests.
* ArgumentTokenizer: Add whitespace tokenizing, refactor lex return type.
* Tag.java: Refactor TAG_VALIDATION_REGEX to accept multi-word tag names.

ewaldhew committed:
[#60] Basic notification add and display window support (#122)

* CoinChangedEvent: Add event to record updated coins
* Rule: Copy over boilerplate from `Coin` classes and add to `Model`, `UI`, `Storage` along with corresponding methods
* NotifyCommand: Add minimal working example
* RuleListPanel, NotificationsWindow: Add to GUI

ewaldhew committed:
v1.4 Documentation (#116)

* DeveloperGuide: Add sections and details to Condition Parser section (#121)
Edit for style, Add sections
Add stylesheet for PDF export
Edit Architecture section
App.F: Revert changes, Add section stub
Update diagrams
* UserGuide: Update for v1.4

ewaldhew committed:
[#105] Basic commands (#118)

* Move prefix checker into multimap itself
* s/MESSAGE_INVALID_COIN_DISPLAYED_INDEX/MESSAGE_INVALID_COMMAND_TARGET/
* CommandTarget: Add new union type wrapper for command targets
Use in `BuyCommand`
Add tests for parsing target
* SellCommand: Add this command
Update to use targets
* AddCommand: Remove unnecessary arguments

* Move `ModelStub` to util class
* BuyCommandParser#parse: Pull out target parsing

ewaldhew committed:
Version 1.4 released (#130)

ewaldhew committed:
Fix help messages for commands (#131)

hotfix

ewaldhew committed:
Fix broken test (#134)

hotfix

eldon-chung committed:
Ui.png: Update to v1.4 (#132)

laichengyu committed:
[#126] Improve SyncCommand (#133)

* Change Set for code list to List
* SyncCommand: Add AsyncHttpClient dependencies
* Add async fetch method
* Change sync command to use async fetch
* Add loading event handlers and loading animation

ewaldhew committed:
Version 1.4 revised (#135)

eldon-chung committed:
TokenType.java: Refactor prefix token type names. (#141)

#146 
* Refactor prefix token type names.

eldon-chung committed:
Coin.java: Refactor get profit, worth and profitability to return Amount (#138)

#146 
* Refactor getProfit, getWorth and getProfitability to return Amount

eldon-chung committed:
CoinBook: Refactor all instances of Address to Coin (#147)

#146 
* CoinBook: Refactor all instances of Address to Coin

eldon-chung committed:
[#46] Add condition generator (#179)

#46, #45, #85, #95
* Add Condition classes to evaluate coins by their metrics and details.
* Add Semantic parser and ConditionGenerator to compose lambdas to evaluate coins
* Implement FindCommand execute
* Add FindCommand test cases

eldon-chung committed:
[#181] Tag#TAG_VALIDATION_REGEX: Only allow single worded tag names (#182)

eldon-chung committed:
[#188] Refactor code condition (#189)

* Code: Add substring evaluation
* CodeCondition: Refactor CodeCondition to evaluate based on substring

eldon-chung committed:
[#184] Coin: Revise equals method to only check Code attribute (#185)

eldon-chung committed:
[#152] CommandTarget#toIndex: Use getSourceIndex method (#183)

This fixes the index determination, which previously did not convert back to the source list and hence always returns the first index in the list, hence only changing the first coin.

eldon-chung committed:
Fix bugs (#193)

Fixes: #166, #191, #192
* AddCommand: Refactor expected prefix from name to code and revise tests
* TagCommand: Refactor Index to CommandTarget to select coins by Code
* Command classes: Refactor COMMAND_ALIASes to reflect the the user guide

eldon-chung committed:
[#194] Add amount condition superclass (#195)

* AmountCondition: Add base class for amount comparison conditions

eldon-chung committed:
[#57] Add news view (#198)

* CoinSubredditList: Add CoinSubredditList to read coin subreddit list
* AddCommand: Add warning message for adding unrecognised coin names
* CoinSubredditList: Add method to get RedditUrl associated with coin
* SubredditNotFound: Add html file to be loaded there is no coin subreddit

ewaldhew committed:
[#109] Coin Amounts fixes (#199)

* Amount: Use `BigDecimal` as backing type
Add arithmetic methods
Update all references
Store in data XML file
Add dynamic scaling of value for display

* Price, CoinBook: Temporarily disable price updating
* Coin: Make new constructor with amount data
Minor code reorganisation
* XmlAdaptedCoin: Add elements to file object
* CollectionUtil: Add utility function to detect missing amount elements
* CoinCard, GuiTestAssert: Fix tests to take amount `String` representation
instead of value representation. The latter should be used only for
storage.
* Amount: Add comments to reflect the above

eldon-chung committed:
ParserUtil: Add warning logging for when Find argument parsing fails (#200)

#196

ewaldhew committed:
[#88] Notification Center (#201)

* Notifications: Add rule types, parsers, and checker
* Rule, NotificationRule: Create class for trigger-action pairs
* XmlAdaptedRule: Store in the rulebook file
* RuleChecker: `LogicManager` component that listens for rule-triggering
events and acts accordingly
* ShowNotificationRequestEvent: Add event to be used for signalling need
for pop-up notification
* ConditionParser: Add conditions for notification rules
* ListNotifsCommand: Add spawn notification list window

eldon-chung committed:
[#167] Refactor tests - Rename addressbook stuff (#202)

* Tests: Rename test coins from person names to actual Coin names
* CoinBookXmls: Rename data in all XML files to contain coin related names
* XmlAdaptedCoinTest: Rename coin related examples
* CodeTest: Rename Code parse examples to coin related names
* ParserUtilTest: Rename test strings to coin related values
* CodeCondition: Rename BTC to BTCZ to load the dummy page for testing
Refactor test method to be case insensitive
* ViewCommandSystemTest: Add test case for selecting index between bounds
* TagCommandSystemTest: Add test case

eldon-chung committed:
[#203] Rename address to coin in all commands, parameter names, and scripts (#204)

#203

ewaldhew committed:
Fix targeted commands (#205)

* TypicalTargets: Rename from `TypicalIndexes`
* ViewCommand, DeleteCommand: Change to targeted

ewaldhew committed:
Deal with unused files (#206)

* Move NewsPanel stuff to unused
* Delete unneeded source files

ewaldhew committed:
Refactor sync command (#208)

ewaldhew committed:
D2 Documentation (#187)

* DeveloperGuide: Update figure content, captions and numbering
Tweak PDF render settings
Various fixes from feedback
* UserGuide: Update command formats to match product
Various fixes from feedback

eldon-chung committed:
[#162] Refactor tests part 2 (#211)

* {Add,Buy,Sell,Find}CommandSystemTest, : Add test case
* {Delete,Tag}CommandTest: Remove test case
* CoinUtil: Add command gets

laichengyu committed:
[#54] Autocomplete feature (#212)

#54 
* Add basic autocomplete
* Add authorship

neilish3re committed:
Update Project Portfolio for neilish3re (#210)

#210 
* Update project portfolio

eldon-chung committed:
SampleDataUtil: Change Coin names to actual Coin names (#215)

#214

eldon-chung committed:
PPP(Eldon-Chung): Add (#213)

eldon-chung committed:
UserGuide: Add examples (#218)

* UserGuide.adoc: Fix example issue in Find command
Add introductory UI section

ewaldhew committed:
[#59] Add generation of coin change events for rule checker (#220)

* CoinChangedEvent: Move diff state data back into `Coin`
* ConditionGenerator: Make separate condition type for diff
* Hook up rule checker to sync procedure
* Add logging messages for matched rules

eldon-chung committed:
SyncCommand: Refactor toString to getAsString during Price construction (#217)

#216

ewaldhew committed:
MainWindow: Make notifications pop-up (#222)

ewaldhew committed:
Add Portfolio Page (ewaldhew) (#190)

* PPP(ewaldhew): Add textual content
* PPP(ewaldhew): Add PR links
* PPP(ewaldhew): Add pull

laichengyu committed:
PPP(laichengyu): Add (#221)

ewaldhew committed:
Version 1.5rc released (#224)

neilish3re committed:
#92 SortCommand feature (#209)

* SortCommand.java: Add related files
Author annotation

laichengyu committed:
Refactor CommandList (#227)

* Refactor CommandList

ewaldhew committed:
MainApp: Fix rulebook init (#228)

The app would always start with an empty rulebook due to this copy-paste error.

laichengyu committed:
Refactor CommandList (#231)

ewaldhew committed:
[#223] Notifications jump to coin (#232)

* Use ControlsFX notifications
* Rule: Refactor Add abstract action command
* Rule: Refactor Move check function out to `RuleChecker`
* RuleChecker#checkAndFire: Also set the extra data
* ShowNotificationRequestEvent: Make this take the index to jump to
* MainWindow#spawnNotification: Set click action to jump to relevant coin
* Notifications: Improve text, Add to User Guide

ewaldhew committed:
NotifyCommand: Fix amount change condition matching (#229)

* NotifyCommand: Fix amount change condition matching
* TokenType: Add the new prefixes into recognized types
* Coin, Price: Actually implement diffing
* ParserUtil: Change log output on parse fail

laichengyu committed:
[#225] Sync data on app startup (#230)

* Sync data on app startup

laichengyu committed:
[#226] Add coin icons to panels (#233)

* Add coin icons to panels

laichengyu committed:
Reduce icon db to 100 items (#239)

* Reduce icon db to 100 items

ewaldhew committed:
Fix Rule Actions (#240)

* ActionCommand: Receive additional information from event
* CoinChangedEvent: Store index of changed coin

laichengyu committed:
[#126] Add loading ui (#241)

* Add loading ui

neilish3re committed:
Update dev guide, user guide and PPP (#234)

laichengyu committed:
Add UrlBuilderTests (#242)

* Add UrlBuilderTests

ewaldhew committed:
Notifications: Add tests (#243)

* Notification Tests
* ModelManagerTest: Add rule tests

ewaldhew committed:
Amount: Disallow negative, Better print (#244)

* Fix negative amounts and better print

laichengyu committed:
Add authorship (#246)

laichengyu committed:
Add FetchUtil tests (#245)

* Add FetchUtil tests

* Refactor FetchUtil and SyncCommand

laichengyu committed:
MainApp: Add tests (#249)

laichengyu committed:
SyncCommand: Add tests (#247)

laichengyu committed:
Remove unused methods (#248)

ewaldhew committed:
Notification Conditions: Fix calculations, Remove useless (#250)

* Remove nonsensical prefixes
* Remove held, bought, sold, made conditions
* WorthChangeCondition, ConditionGeneratorTest: Fix calculations

laichengyu committed:
Attempt to fix coveralls badge (#253)

laichengyu committed:
Edit Async fetch comments (#251)

* Edit asyncFetch comments and javadoc

ewaldhew committed:
[#57] Charts Panel (#256)

#57 
* SyncCommand: Get historical prices
* SyncCommand: Factor out fetch procedures
* Price: Add lists to store historical data
* ChartsPanel: Add to app
* ChartsPanel: Add processing for coin data
* MainWindow: Add new panel
* Author annotation

laichengyu committed:
UGDG/PPP(laichengyu): Update (#259)

* Update UserGuide
* Update PPP
* Update autocomplete in DG
* Update dev guide

laichengyu committed:
DevGuide update (#261)

* Update UserGuide
* Update PPP
* Update DeveloperGuide

ewaldhew committed:
Documentation updates (Notifications) (#260)

* Edit UGDG
* Edit Diagrams
* Notifications and various doc fixes

neilish3re committed:
Update User Guide, Dev Guide and PPP (#257)

laichengyu committed:
Fix UG typos (#264)

* Fix UG typos

ewaldhew committed:
Beautify App (#263)

* Notifications: Add coin icon
* IconUtil: Move icon getter to this file
* CoinListCard: Move amount display to right edge
* StatusBarFooter: Add filter status
* MainWindow: Reset filter before notification coin jump
* DarkTheme: Tweak colors
* StatusBarFooter: Make new element for filter

laichengyu committed:
Include sync feature in DG (#266)

ewaldhew committed:
Final edits (#267)

* UserGuide: Update
* DeveloperGuide: Fix figure numbers
* UserGuide: Warn about find filtering
* UserGuide: Use basic xref label
* Final edits

ewaldhew committed:
SyncCommand: Prevent crash when internet connection out (#268)

ewaldhew committed:
Final PPP (#269)

ewaldhew committed:
Fill in final missing documentation (#270)

ewaldhew committed:
Version 1.5 released (#271)
